File: hasPath_bFS.py
"""
Input 1: a maze represented by a 2D array
0 0 1 0 0
0 0 0 0 0
0 0 0 1 0
1 1 0 1 1
0 0 0 0 0
Input 2: start coordinate (rowStart, colStart) = (0, 4)
Input 3: destination coordinate (rowDest, colDest) = (4, 4)
Output: true
Explanation: One possible way is : left -> down -> left -> down -> right -> down -> right.

Time complexity O(MN)
SC is O(MN) in queue

"""
class Solution:
    def hasPath(self,maze: List[List[int]],start: List[int],destination: List[int])->bool:
        if maze is None or len(maze)==0:
            return False 
        m=len(maze)
        n=len(maze[0])
        q=[]
        q.append(start)
        dirs=[[1,0],[0,-1],[0,1],[-1,0]]
        while q:
            curr=q.pop()
            if(curr[0]==destination[0] and curr[1]==destination[1]):
                return True 
            for dir in dirs:
                i=curr[0]
                j=curr[1]
                while(i>=0 and j>=0 and i<m and j<n and maze[i][j]!=1):
                    i=i+dir[0]
                    j=j+dir[1]
                i=i-dir[0]
                j=j-dir[1]
                if(maze[i][j]!=2):
                    q.append([i,j])
                    maze[i][j]=2
        return False
        

File: hasPath_BFS.cpp
// Time Complexity : O((m * n) * max(m, n)) 
// Space Complexity : O(m * n)
// Did this code successfully run on Leetcode : yes

// Three line explanation of solution in plain english
/* Perform BFS operation, change the direction only when the boundry or wall is hit.
 * To indicate visited cell, set its value to 2. 
 */

class Solution {
public:
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination) {

        if (maze.size() == 0)
        {
            return false;
        }

        int m = maze.size();
        int n = maze[0].size();

        vector<vector<int>> dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        queue<vector<int>> q;

        maze[start[0]][start[1]] = 2;
        q.push({start[0], start[1]});

        while (!q.empty())
        {
            int currRow = q.front()[0];
            int currCol = q.front()[1];
            q.pop();

            for (const auto& dir : dirs)
            {
                int i = currRow;
                int j = currCol;

                while (i < m && i >= 0 && j < n && j >= 0 && maze[i][j] != 1)
                {
                    i += dir[0];
                    j += dir[1];
                }

                i -= dir[0];
                j -= dir[1];

                if (i == destination[0] && j == destination[1])
                {
                    return true;
                }

                if (maze[i][j] == 0)
                {
                    maze[i][j] = 2;
                    q.push({i, j});
                }
            }
        }

        return false;
    }
};